hermes-consumers/src/test/groovy/pl/allegro/tech/hermes/consumers/supervisor/process/ConsumerProcessSupervisorSignalsTest.groovy

```diff
@@ -0,0 +1,160 @@
+    }
+
+    def "should restart unhealthy process"() {
+        given:
+        def consumer1 = Stub(Consumer)
+        def consumer2 = Stub(Consumer)
+        def consumers = [consumer1, consumer2]
+
+        def conditions = new PollingConditions(timeout: 2)
+        def createConsumerFactoryMethodCallsCount = 0
+
+        consumer1.consume(_) >> {
+            Runnable signalsInterrupt -> signalsInterrupt.run()
+        }
+        consumer2.consume(_) >> {
+            Runnable signalsInterrupt -> signalsInterrupt.run()
+        }
+
+        consumerFactory.createConsumer(subscription) >>
+                { subscription -> consumers[createConsumerFactoryMethodCallsCount++]}
+
+        consumerProcessSupervisor.accept(
+                Signal.of(Signal.SignalType.START, subscription.getQualifiedName(), subscription))
+        consumerProcessSupervisor.run()
+
+        when:
+        consumerProcessSupervisor.accept(
+                Signal.of(Signal.SignalType.RESTART_UNHEALTHY, subscription.getQualifiedName()))
+
+        then:
+        conditions.eventually {
+            consumerProcessSupervisor.run()
+            assert createConsumerFactoryMethodCallsCount == 2
+            assert consumerProcessSupervisor.countRunningProcesses() == 1
+        }
+    }
+
+    @Unroll
+    def "should #signalName process"() {
+        given:
+        def delay = 0.5
+        def timestampContainer = new TimestampContainer()
+
+        def consumer = Stub(Consumer)
+        def conditions = new PollingConditions(timeout: 4 * delay, delay: delay)
+
+        consumer.consume(_) >> { Runnable signalsInterrupt ->
+            signalsInterrupt.run()
+            timestampContainer.update()
+        }
+
+        consumerFactory.createConsumer(subscription) >> consumer
+        consumerProcessSupervisor.accept(
+                Signal.of(Signal.SignalType.START, subscription.getQualifiedName(), subscription))
+        consumerProcessSupervisor.run()
+
+        when:
+        consumerProcessSupervisor.accept(
+                Signal.of(signalType, subscription.getQualifiedName(), subscription))
+        consumerProcessSupervisor.run()
+
+        then:
+        conditions.within(4 * delay) {
+            assert System.currentTimeMillis() - timestampContainer.timestamp > 2 * delay * 1000
+            assert consumerProcessSupervisor.countRunningProcesses() == 0
+        }
+
+        where:
+        signalType             |   signalName
+        Signal.SignalType.STOP |   "stop"
+        Signal.SignalType.KILL |   "kill"
+    }
+
+
+    def "should force kill stopped process if it is running to long"() {
+        given:
+        def delay = 0.5
+        def timestampContainer = new TimestampContainer()
+
+        def consumer = Stub(Consumer)
+        def conditions = new PollingConditions(timeout: 4 * delay, delay: delay)
+
+        consumer.consume(_) >> { Runnable signalsInterrupt ->
+            while (true) {
+                Thread.sleep(200)
```

Comment:
Unfortunatelly we need this sleep() for consumer to be stoppable. Without this java does not kill the process (but FutureTask says that cancel() method was stopped correctly). This can be good way to simulate the real production problem and find some better solution to kill stale processes.